New planner

dodomorandi opened the pull request:
I implemented a new system to handle the TransferPlanner.

Now the plan stay "fixed" instead of moving with the player. It is also possible to move the point where the plan will "start" along the current orbit. This should create a new way the game can be played, in a more KSP-oriented way.

I have got lots of ideas in my mind, just give me some feedback about this modifications. If you think that it could be a good starting point, I will probably continue to develop these features!

Just a few images to give an idea of what has been done:
![screenshot01](https://cloud.githubusercontent.com/assets/350168/8766956/42f06a2e-2e4c-11e5-9973-e2c49c116d0e.png)
You can notice other buttons in the top right. They are related to the maneuver time.

![screenshot02](https://cloud.githubusercontent.com/assets/350168/8766957/42f28156-2e4c-11e5-9ca2-ef9b5b837ef1.png)
The central button, as always, resets the value. In this case, it will move the maneuver starting point to the actual position.

![screenshot03](https://cloud.githubusercontent.com/assets/350168/8766959/42fb183e-2e4c-11e5-85cc-8c30843b74a3.png)
You can easily plan the maneuver just moving the time in the future

![screenshot04](https://cloud.githubusercontent.com/assets/350168/8766958/42f78034-2e4c-11e5-9b80-e936bc24da45.png)
The blue pointer works as always, but the delta-delta-V is also displayed. For now, only the resulting value is displayed instead of prograde+normal+rotational. Maybe in the future.

![screenshot05](https://cloud.githubusercontent.com/assets/350168/8766960/42fc2422-2e4c-11e5-9e58-3a2eb8124e60.png)
Almost there... it is time to start the maneuver!

![screenshot06](https://cloud.githubusercontent.com/assets/350168/8766961/43018e80-2e4c-11e5-9064-a7e02232db44.png)
Indeed, the blue pointer is where we should accelerate to obtain the resulting orbit. It is difficult to reach the 0m/s value, but generally reaching a low value still gives a very good result!

Give it a try!

nozmajner commented:
Nice! 
I think I'd use a hourglass icon for it in the button though. But I can replace that easily later.
Does it work with the time forward/backward functin we have in the map?

dodomorandi commented:
@nozmajner Very nice idea! I will committente an hourglass ASAP!

What do you mean about the time controllers? I mean, the plan is indipendent from the current position\velocity. If you mean about creating an alter-ego of the ship on the new orbit... It would be a nice idea. However it is almost useless until multiple planners are supported, IMHO.

nozmajner commented:
Yupp, I think it represents it better. You know I could have done that for you, that's what artists are for. :)

I'm refering to those time controllers on the lower right in orbit map, which you can use to see where the bodies and crafts will be at any given time in the past or the future. It helps planning your rendez-vous. 

![screenshot-20150719-221115](https://cloud.githubusercontent.com/assets/4182678/8767647/3d62ee8a-2e63-11e5-90a9-dbc3d8c7e85d.png)

I think it would be uselful to have that alter-ego on the new orbit. You could use it to see if you actually get any close to your intended destination on that orbit, or you will at the intersecting point hours after it.

dodomorandi commented:
@nozmajner Whoops, I just noticed that, from my post, it seems that I was not aware of the time controls ;). I was just wondering how to make them interact with planning system... And I completely forgot about rendez-vous problems! You are absolutely right: this feature are very important, and I will implement it ASAP, probably tomorrow.
However, IMHO time controls need to be changed a bit: it is almost impossible to accurately control the time dimension... Now I will focus on planner, then I will probably open another branch to propose a new system.

nozmajner commented:
Yes, they have a quite big timestep even on the lowest setting.

impaktor commented:
wow, that's a lot of activity on this PR in just a few hours. @nozmajner looking at the timestamps, I'm worried about your sleeping pattern.

@dodomorandi :+1: nice to see some work on this screen.

Maybe previous contributor to this screen and KSP-plaer @dvolk is interested in this PR.

impaktor commented:
@dodomorandi Looking at the code, you only need to add the translation string to en.json, the other languages will be added by the translation system/bot after it has been merged.

I'm not sure what happens if we merge non-english json translation files that have been changed? Possibly we get a conflict, or possibly they will just be overwritten.

dodomorandi commented:
@impaktor Sorry, I thought it worked in a sort of mixed automatic-manual way. It is possible to perform a translation test or it is better to revert back changes?

dodomorandi commented:
Developers and artists, I need your opinion about a couple of things. I will wait for your replies before committing and pushing any code (I just do not want to mess around!).
I added the support of the time control over the orbital maneuver, but I started having some troubles with... too many colored squares! I decided to make some graphical changes, tell me what do you think about them. Oh, do not worry about the quality of the images/icons: they are almost placeholders, just a bit nicer than a white square ;)

![screenshot1](https://cloud.githubusercontent.com/assets/350168/8787616/650cca96-2f35-11e5-88c9-423e00f53767.png)
In this image you will notice that the position of the ship is... a ship. Next to the hourglass it is shown the word "Now". Do not bother about the positioning of this label, I will talk about this very soon...

![screenshot2](https://cloud.githubusercontent.com/assets/350168/8787512/eb0eda36-2f34-11e5-8437-ce3aab38aa62.png)
Here the orbit. Now the start of the orbit maneuver is an ugly blue arrow, but maybe it is still better than a blue square. You can notice that next to the hourglass is shown the time the maneuver has to start, but there is not enough space... Shall I move everything a bit on the left or you have some better ideas?

![screenshot3](https://cloud.githubusercontent.com/assets/350168/8787513/eb10506e-2f34-11e5-8ed8-9d644c00f625.png)
Ok, now we are using the time controls. As you can see I decided to keep the actual position on the screen and draw the "future" position, just colouring the ship in red. What do you think about it?

![screenshot4](https://cloud.githubusercontent.com/assets/350168/8787514/eb1108b0-2f34-11e5-8512-ea8e3176ce05.png)
And here we go! Same situation, another ship icon!

Give me some feedback, it is really appreciated! I do not want to ruin the old-style-but-fancy-looking vision of the game, but maybe some eye-candy could be ok.
I will wait for your opinions!

impaktor commented:
> or it is better to revert back changes?

Yes, this is what I would prefer.  I.e. just `en.json` changed, and the others reverted, especially since our Transifex expert is on holiday. We can squash such a "fixup commit" with the other one before merge, to have the commit history clean.

Using a ship icon does (at least in this example) give it a more "gamey" feel and less "sciency", I think, and that could be bad or good, depending on how you look at it. 

I have been thinking that Pioneer needs to be more self documenting, and especially this orbital screen is quite "advanced" and not so intuitive. My thinking was that one could perhaps have a legend with the icons and descriptive text (unless that is already provided by the mouse over, been a while since I used this screen).

Also, if it is possible to have dashed and/or dotted lines, that makes things more easy to separate. I'd guess up to 10% of our players (and 100% of our current one artists) might be colour blind (green-red blindness being quite common among men).

Also, thanks for all the screenshots and descriptions, makes it easier to see what is going on.

> Shall I move everything a bit on the left or you have some better ideas?

Move whatever you need. (There's also space to the lower left, downside with that is we might get a bit "bottom heavy" with buttons.)

dodomorandi commented:
Ok, I fixed up the commit and fast-forwarded.

The dotted lines are a great idea! I like it! I already wanted to change the behaviour of the orbit representation (IMHO the resolution should scale appropriately with the zoom level), adding a support for dotted ellipses would be nice as well. I see what i can do!

I also already planned to write down a mouse over support. My idea is to give information about the hovered position of the orbit (height from the focal point/orbiting point, speed, eccentricity and so on) and also of a "interesting point" like the position of the ship, the peri/apohelion and whatever could be useful... But I need to work on it.

These changes will probably need some time to be written, therefore... What could be the best option for now? Shall I put the "gamey" icons (I loved this expression!) or the original squares? Maybe I could also use a triangle or anything else to represent the maneuver starting point... As always, give me your feedback! Oh, thank you @impaktor, your suggestions are great!

Meanwhile I will try to place these buttons in better places!

impaktor commented:
Perhaps you'd be interested in commenting back line 34 in `data/ui/InfoView.lua`. I'm not sure if that was a first stab at moving the orbital screen to Lua (all the blue in-game windows), but it does have some info.

Another thing I'd like to see is replace these ugly white blobs of planets with just a spherical green grid. I know there have been requests to see terrain features as well, but that is way beyond the current state of the code, and might "spoil" the surprise when going to the planet, from a game perspective.

laarmen commented:
Just passing through, but if you want us (well, them) to look at some
experimental code, you can always just put it on a separate branch and tell
us about it. That way you don't mess up your current PR but we can still
make in-line remarks, and, if you're lucky, even test the thing ;-)

Le mar. 21 juil. 2015 à 10:02, Karl F <notifications@github.com> a écrit :

> Perhaps you'd be interested in commenting back line 34 in
> data/ui/InfoView.lua. I'm not sure if that was a first stab at moving the
> orbital screen to Lua (all the blue in-game windows), but it does have some
> info.
>
> Another thing I'd like to see is replace these ugly white blobs of planets
> with just a spherical green grid. I know there have been requests to see
> terrain features as well, but that is way beyond the current state of the
> code, and might "spoil" the surprise when going to the planet, from a game
> perspective.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pioneerspacesim/pioneer/pull/3426#issuecomment-123206195>
> .
>

fluffyfreak commented:
Maybe do those `ugly white blobs of planets` in another PR I think this one is busy enough :)

The screenshots look great though and the code looks good.

laarmen commented:
Oh, I actually meant just pushing a branch on your repo and drop the name
in a comment in this PR, not creating a whole new one :)

On Tue, Jul 21, 2015 at 2:55 PM Edoardo Morandi <notifications@github.com>
wrote:

> Thank you guys for your feedback! As @laarmen <https://github.com/laarmen>
> told me I pushed a new further branch to let you look at the code for new
> time system and icons. Obviously, imaged can be easily replaced with other
> (nicer) content or with hard-coded shapes. The PR is #3427
> <https://github.com/pioneerspacesim/pioneer/pull/3427>.
> @impaktor <https://github.com/impaktor> That tab is quite nice! Maybe
> some of these information could be put inside a box popping up on orbitals
> mouseover.
> @impaktor <https://github.com/impaktor> @fluffyfreak
> <https://github.com/fluffyfreak> I will start working on a new UI ASAP,
> but I think that some gameplay-related features should have priority.
>
> P.s.: I will post a fix for speed calculation on non-elliptic orbit soon,
> I already noticed the bug.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pioneerspacesim/pioneer/pull/3426#issuecomment-123296337>
> .
>

dodomorandi commented:
Whoops! Sorry! Closed! The branch is at https://github.com/dodomorandi/pioneer/tree/new_planner_with_icons

laarmen commented:
No problem at all.

impaktor commented on AUTHORS.txt:
```diff
@@ -32,6 +32,7 @@ The Pioneer Developers are:
 * Dionisis Kontominas
 * Dmitry Marakasov
+* Edoardo Morandi
 * Emery Hemingway
```
You are already in authors.

impaktor commented:
I'm confused looking at the "Files changed", as it includes diffs to things we already have in master, like fluffyfreak's AUTHOR edit, I just (unnecessarily) commented on. I though github was smarter than this? Or put it in a shorter way: why are there commits in this PR that are already in master?

dodomorandi commented:
I think it is a bit my fault: I wrote a PR for fixing labels, then I started working on these changes. Next day that PR has been accepted and I rebased this PR on the new master. When I squashed and fast-forwarded everything to remove the lines on json files, GitHub showed commits that have been picked... Yes, it sounds weird, but it seems to be what really happened! :)

nozmajner commented:
Looking good.
Don't worry about the icons much, I can replace them down the line, or while you're working on the planner, if it needs more work. (I can work on them in the weekend, It's bikehike time for me until that). Don't give them arbitrary scaling please, I will set their size in the icon.

The maneuver time display could be most easily solved in my opinion if the whole planner thing is moved to the left side of the screen. But one question: is it even needed to be that detailed? an ETA countdown would alleviate that problem if you aren't planning months or years ahead. And it can still be a fixed time, only the display would count down, and a tool-tip could show the exact date and time.

I think the dashed line solution would be nice, not just because of color-blindness, but in general. The more contrast, the clear the differentiation. A good thing to remember is that color information is the last thing our brain notices, and shape is much higher on this order.

One other thing that would be useful is an indication of the orbit direction. the mock-up under the bellow link I solved it by having the ship position indicator triangle pointing in the orbit direction (not sure how difficult to implement it), and also fading the orbit line at the end of the circle. I think @dvolk had a similar experiment a while ago.

And I can second what @impaktor is saying about having a realistic ground display on the orbital map. It would clutter up the map with not really important information in my opinion. A separate ground track/map display would be more useful for that.

Also I made some mock-ups about the orbital map a few years agou. [Might give you some ideas.](http://pioneerspacesim.net/forum/viewtopic.php?f=3&t=24&)
![orbit planner](https://lh3.googleusercontent.com/-u9Iysx4mVjI/UXUu88ZVHcI/AAAAAAAABcU/zTXHlFKjKC8/s2048-Ic42/sysmap_loworbit2.jp)
![legend](https://lh3.googleusercontent.com/-xNLXjZx402g/UXUu8wgcbiI/AAAAAAAABdE/LJlfOcfXyj8/s2048-Ic42/sysmap_loworbit2_expl.jpg)

dodomorandi commented:
@nozmajner Whooo!!! Holy.... it is totally awesome! I will probably "steal" some useful ideas from your mock-up! :D And also the other ideas are great! I already thought about rotating the ship indicator (whatever it could be), but I need to think about the best solution (I do not know if it could be better to project the velocity vector on the screen or something directly calculated from the projected ellipse).

Just one question. I used a customizable size for icons because the Gui::TexturedQuad does not retain any sort of size information. Do you think it could be better to implement a property and its access method for Gui::TexturedQuad or saving the sizes (taken from the Graphics::TextureBuilder?) of the icons?

IMHO the problem with the "ugly white blobs" is that they are almost shadeless, without any kind of surrounding projections and so on. In your mock-up, they appear completely different, and I think some sort of similar results would be the best! I still need to hack some code to completely understand how to obtain best results with the GUI. ;)

Actually, I already changed the "date system" with a sort of ETA. I am working on the mouse over feature, it will probably take me a bit to find the best solution. I will commit the changes on the other branch ASAP and I will post some screenshot down here, as always!

impaktor commented:
I've noticed that none of these features of orbital planning are described in the manual [wiki](http://pioneerwiki.com/wiki/Manual), perhaps when you need a break from coding, you could add a new screenshot, and some text for Orbital Planner, if not, I will ask someone else who wants to contribute. No worries.

> fading the orbit line at the end of the circle. I think @dvolk had a similar experiment a while ago.

I had forgotten about that. Check out the screenshot in #1421.

Also, worth keeping in mind, as fluffyfreak mentioned, pull request that grow in extent are difficult to review, and take longer to merge. No-brainers can be merged quickly.

fluffyfreak commented:
> Just one question. I used a customizable size for icons because the Gui::TexturedQuad does not retain any sort of size information. Do you think it could be better to implement a property and its access method for Gui::TexturedQuad or saving the sizes (taken from the Graphics::TextureBuilder?) of the icons?

I'm not sure I understand what you're saying here. Do you mean that you need to be able to resize a `Gui::TextureQuad`?

dodomorandi commented:
I do not know if I misunderstood @nozmajner when he said 'Don't give them arbitrary scaling please, I will set their size in the icon.', because I set the size of the images inside the new SystemView::Draw, scaling the images. My problem is that I do not have any kind of information about the original size, and I have to pass a 2d vector to the Gui::TexturedQuad::draw.
I would completely agree that a Gui::TexturedQuad could be drawn just using the original size (artists do not have to fix the code in this way), but I was wandering what the best solution could be. Now I have a simpler idea: if the 'pos' parameter of the Gui::TexturedQuad::draw method is (0,0), the 'texSize' could be used. What do you think about it?

fluffyfreak commented:
I will have a look at the code to try and figure this out, but if you just pass `Gui::TextureQuad` a texture then it uses the textures dimensions for rendering. They're only meant for screen quads, icons/buttons/etc for example that are not used in a 3D environment.

One thing I have noticed is that the code won't compile for Visual Studio 2013 which our windows developers (including me) use. This is because VS2013 is lagging behind the standards and doesn't support the use of `and`, `not`, `or` so you need to use the C style `&&`, `!`, `||` instead.

dodomorandi commented:
About the Gui::TexturedQuad... that is what I also expected, but look at the declaration:

    class TexturedQuad {
    public:
        TexturedQuad(Graphics::Renderer *r, Graphics::Texture *texture, const vector2f &pos, const vector2f &size, RenderState *state);
         ...
     }

The size parameter is not optional, and inside the definition there is not any sort of control about the value of the vector. Indeed, I was proposing to change the behaviour of the method in a way that if size is equal to vector2f(0.f), then the size of the object will be texture->GetDescriptor().texSize. What do you think about it? :)

About the problem with VS2013... Sorry guys, as a Linux developer, i REALLY did not expect M$ not completely using C++98 standards as default... Crap, it is nearly 20 years (and they have no excuses: C++ is not C, if some code breaks, it need to be fixed)... I will fix the commits ASAP. What do you think if I add the flag "-fno-operator-names" to the compiler? In this way Linux developers have an easy way to detect these kind of incompatibilities. Or, on the contrary, we could pass "/Za" to VS... I could open a PR for that, it is a fast and simple change, and I think that it will not give any sort of problems...

fluffyfreak commented:
Ah, size should really be called "scale". it's not the dimensions of the object but how much it gets scaled by onscreen. You can safely leave it at `vector2f(1.f, 1.f)` if you want.

The whole `Gui::TextureQuad` and other parts of the old-GUI we would like to eventually remove but these screens are still littered with them so they're not quite right in places.

This means that no-one owns them in a sense so you can do what you need too with them :)

The other issue with visual studio, don't worry about it, I will be pushing us onto VS2015 at some point in the coming months and will evaluate it then.

dodomorandi commented:
You just told me an interesting thing: is it possible for me to use the elements of the new GUI or everything inside SystemView have to be changed? I am asking because it could be a bad idea to make a new SystemView GUI with an obsolete system.
If you think that it could be possible to gradually migrate to the new system (and I think I will need a hint about where to search in the code), I can just start using the new GUI and, time by time, update the rest. Otherwise, if this is not possible, it could be better to stop implementing new features in this PR, to start the migration in another one and then to continue the implementation of new features.

fluffyfreak commented:
Don't worry about it being obsolete, it's been that way for about 2 years :)
I've started rewriting it twice, and @robn has done the same.
Any work you do here would be implemented in a new system if one was ever written so no work is wasted or lost :)

For now, with this PR, carry on as you have started.

dodomorandi commented:
@fluffyfreak Mmmmhh... Are you sure that the "size" parameter represents the scale factor? I set it to vector2f(1.f) and now icons are invisible, probably 1x1 pixels. I also checked inside graphics/Drawables.cpp, it seems to be treated like a size, not a scale:

    TexturedQuad::TexturedQuad(Graphics::Renderer *r, Graphics::Texture *texture, const vector2f &pos, const vector2f &size, RenderState *state)
        : m_texture(RefCountedPtr<Graphics::Texture>(texture))
    {
        ...
        const vector2f texSize = m_texture->GetDescriptor().texSize;

        vertices.Add(vector3f(pos.x,        pos.y,        0.0f), vector2f(texPos.x,           texPos.y+texSize.y));
        vertices.Add(vector3f(pos.x,        pos.y+size.y, 0.0f), vector2f(texPos.x,           texPos.y));
        vertices.Add(vector3f(pos.x+size.x, pos.y,        0.0f), vector2f(texPos.x+texSize.x, texPos.y+texSize.y));
        vertices.Add(vector3f(pos.x+size.x, pos.y+size.y, 0.0f), vector2f(texPos.x+texSize.x, texPos.y));
        ...
        m_vertexBuffer->Populate(vertices);
    }

Maybe something have been changed for some reason. Honestly, I would prefer to have a scaling factor instead of a size factor, for obvious reasons. If you agree, I can just change from "size" to "scale" and replace the "pos.X + size.X" (with X equal to x or y) to "pos.X + texSize.X * scale.X".

fluffyfreak commented:
hmmm, you've got me I don't know :)
Although I've refactored that code I didn't write it originally and everytime I change it something else gets broken.

If you want to change it then go right ahead just be sure to check the other places that it's used.

nozmajner commented:
Sorry, I might had confused something. I was assuming that icons always get scaled to a certain size or something, and they always get pixelated regardless of image size. In general, not here specifically.

dodomorandi commented:
@nozmajner @fluffyfreak Guys, I have problems to remember to code I wrote in the past... You are great at handling a OS project like this...

I almost finished writing some interesting changes, I hope I can post something later on. Then I will also give a look to TexturedQuad, where it is referenced and it is a good or a bad idea to change the default behaviour...

dodomorandi commented:
I made some new changes in the separate branch (https://github.com/dodomorandi/pioneer/tree/new_planner_with_icons). Not everything is working smooth as I wanted (I put some details inside the commit), but it seems to be almost usable and playable. Tomorrow I will add some nice screenshots, now it is really time to go to bed ;)

Meanwhile, if you want to test these new features, feedback are welcome!

dodomorandi commented:
Some good and bad news... Here the good ones:
![schermata da 2015-07-23 21-15-19](https://cloud.githubusercontent.com/assets/350168/8859818/09f0fb44-3180-11e5-8980-49e30dcb8751.png)
As you can see, an initial and partially working system for mouse detection has been implemented. However, the code just saves the 2d points evaluated by PutOrbit and use them to track the orbit and the 3d position. A very sub-sub-(sub-sub)-optimal solution.

Now the bad news. A behaviour I did not expect:
![schermata da 2015-07-23 21-13-58](https://cloud.githubusercontent.com/assets/350168/8859877/6e7c81f0-3180-11e5-8ecd-c007dad287f4.png)
In fullscreen mode, at 1366x768, the calculation of the 2d position of the orbital points using Gui::Screen::Project does not work correctly... and I do not know why.

If you want, play with the branch, hack the code, give me feedback. Meanwhile, I will probably change completely the tracking system, trying to find the algebraic solution for the projection of the ellipse (which, in turn, is an ellipse), the parabola and the hyperbole (for which I do not have the minimum idea... I will need to study).

I hope I can post some news soon!

impaktor commented:
If the player wants ETA at some point in orbit, could they not just fast forward to that position? So a simpler solution would be to have the fast forward/backward (lower right corner) show the date and time difference?

I'm just trying to save you a lot of headache.

dodomorandi commented:
Thank you for the suggestion, but I want to explain the reason I would like to blow my mind :)
IMHO, it would be awesome to have 4D information about a specific point of an orbit (altitude, speed, ETA related to your position, the maneuver) and whatever comes in mind. And, in theory, this could be also done using the current system, even if the mouseover is more user friendly...
But what if I would like to drag the maneuver starting point? Obviously a system to analyse the 2d position of the cursor in relationship of the orbits in the screen is needed. But yeah, it is still possible to do this using button controls.
Now, think about the possibility of multiple planners. The easiest solution would be something like "Ctrl+click on orbit -> new transfer plan created on that point", and I thing that handling a system with multiple planner controls would be very hard, especially from the user point of view.

Said that, I know that it will not be easy, but I think that adding this kind of support would be very helpful for future features. Obviously, I am trying to write the basics and, later I would have focused on "eye-candy" features (i.e. the drag system) and advanced ones (the multiple planning system support).

However, I really want to consider your opinion because you have a better global idea of the project, the priorities and the direction it should take. If you think I should focus on different things, no problem at all! :)

impaktor commented:
> However, I really want to consider your opinion because you have a better global idea of the project, the priorities and the direction it should take.

I was just thinking out loud, to see if there was an easier way to go about it. Keep doing what you do, and we see what comes out of it. I'm not against adding new cool stuff.

dodomorandi commented:
Hello everybody!
I decided to put some of the new features I coded in this PR. Obviously, I left eye-candies in the other branch, and I only added things related to orbit transfer planner.
The main reason of this decision has been mainly related to future development and KISS-rules to make life easier to @impaktor, @nozmajner, @fluffyfreak and anyone else that would like to review the commits. Now it should be a bit easier to track the real changes.
About some details... you will find in these commits almost the same changes I saw in the screenshots about the other branch... but without eye-candies and cursor detection (which I removed from the other branch as well, I will implement a much better solution).
We could start thinking about what should be made in this PR: as you wisely said, it is better to keep PRs as simple as possible. If you think that there are enough changes, we can close it and I will start with the implementation of other features elsewhere. Otherwise, as always, just write down you feedback, I will be happy to develop everything in the best way I can!

nozmajner commented:
I can't do much reviewing myself, because I'm yet to learn how to compile the game. 
But I can start working on some icons in the meantime. Which ones are needed specifically? I saw one representing the ship, and you already have the hourglass. And one indicating the maneuver on the path.

dodomorandi commented:
@nozmajner Mmmhhh... I can only help you if you have a Linux OS. I do not think that trying to cross compile would be a nice idea. If someone did it, let me now! ;)
About icons: IMHO the best option would be to add them in another PR, and probably we should create it only when this one is closed. Obviously, we can make some tests using my other branch, but you will need a working executable... You can ask someone like @impaktor and @fluffyfreak if they could build an experimental package (but I have to make the change to use the naive size of the images).
I was thinking about some icons...
* A nicer ship. The best option would be to have different symbols for present and future ships, for current and planned orbit.
* A much much nicer starting maneuver symbol. Probably, even a circle would be nicer, like in your mock-up
* A better hourglass
* More intuitive periapsis and apoapsis
* Ascending and descending nodes. I am planning to show these, but I need the orbital mouse detection system, and I am working on it. If you want, you can start thinking about it. Wikipedia, at https://en.wikipedia.org/wiki/Orbital_node, shows two symbols for these two node. Maybe it could be a good staring point.
* IMHO, we need different symbols for prograde, anti-prograde, radial, anti-radial, normal, anti-normal and burn direction in first and third person mode (I completely agree about not using the same icon with different colors).

If you have got other ideas, they are welcome!

nozmajner commented:
How about these icons and reticules?
![ui_orbitmap_icons01](https://cloud.githubusercontent.com/assets/4182678/8947497/d8c100ee-359b-11e5-9fd0-ed413a5e8837.png)
* Ship is indicated by a Box
* Orbit tail end is faded to indicate orbit direction
* Node is indicated with dot and circle
* Those triangle apsis icons were bugging me too. I changed it to something similar KSP has, whatitsname, boxtriangle. Outline only for Apoapsis, and it's pointing up, Filled pointing down for Periapsis. They have a line trough the sphere (not sure if it's easy to implement, but thought it's useful. Maybe later)
* Ascending and descending nodes. I re-purposed the triangles for them.
* __+__ for future ship position, visible both on current and planned orbit.
* I've also included an equator line, it makes visualizing easier.

I've also included the reticules for 1st and 3rd person views:
* Same as now for pro and retrograde
* Up and down pointing triangle for normal and antinormal
* Diverging and converging triangles for radials
* Same dot and circle as on the map for the Maneuver node, with additional deltaV display and bar, and ETA display. (these would be nice in my opinion)
* I've included an 'anti-maneuver' too, could be helpful in some cases I guess, like when you are using rear view but want to keep on the node.

No need to help, I will pick compiling up sometimes I guess. And the official windows builds are cross-compiled on linux if I'm correct.

walterar commented:
Within this package you have what you need, to try this PR
https://dl.dropboxusercontent.com/u/15391580/Pioneer/3426-L64-W32.zip

modelcompiler-3426.exe and pioneer-3426.exe for Win 32
modelcompiler-3426 and pioneer-3426 for Linux 64

You can copy directly to the Pioneer directory and run independently.

en.json and orbit_start_big.png are in data-3426-mod.zip

You can put it in the mods folder, or decompress directly into the data folder

You will need:
http://sourceforge.net/projects/pioneerspacesim/files/pioneer-20150715-win32.7z/download
or
http://sourceforge.net/projects/pioneerspacesim/files/pioneer-20150715-linux64.tar.bz2/download

I hope you understand, my English is null :)

dodomorandi commented:
@nozmajner Very nice work! Only one little objection, mainly programming-related: what do you think of having the projection of an orbit on the planet instead of the equator line? I think that it could help in the same way but it also helps the understanding of the orbits.
You seem almost ready, aren't you? I can create a new PR and start implementing changes. It should be quite fast, even because you already worked on many things!

nozmajner commented:
@dodomorandi I think they aren't mutually exclusive. On my first mockups I made the things below the equator darker, and I also did it on this one at first. but I figured it might be confusing with day-night side, that's why I just added an equator line. Maybe a tinted south hemisphere with orbits projected on the surface would work.
@walterar Thank you very much.

nozmajner commented:
@dodomorandi I've made the icons that were in your planner with icons branch.
https://www.dropbox.com/s/hmgzis2ojbsxjuz/planner_icons.zip?dl=0

I've replaced the apsis icons too, hopefully they are more intuitive this way.
I've also replaced the hourglass icon with a simple clock. I was thinking about it, and the hourglass is more like a timer, like for burntime, that's the reason.
![icons](https://cloud.githubusercontent.com/assets/4182678/9035286/ca63b6f6-39d6-11e5-82f9-9e45c8ed00aa.png)

I've already gave them to @impaktor , he can put them in when he merges your branch.

impaktor commented:
Yes, I'll be trying this out before merge next time I'm at a capable enough computer. Might be today, might be later.

impaktor commented:
Played around some in the F2->F6 menu, then hyperjump -> segfault!

```
onLeaveSystem:total:26,removed:26
0 clouds brought over
Started hyperspacing...

Program received signal SIGSEGV, Segmentation fault.
SystemBody::GetSuperType (this=this@entry=0x0) at StarSystem.cpp:192
192switch (m_type) {
(gdb) bt
#0  SystemBody::GetSuperType (this=this@entry=0x0) at StarSystem.cpp:192
#1  0x0000000000432a4e in GetMass (this=0x0) at galaxy/StarSystem.h:128
#2  DynamicBody::ComputeOrbit (this=0x1f8d9340) at DynamicBody.cpp:304
#3  0x000000000057f864 in WorldView::UpdateProjectedObjects (this=this@entry=0x200f2e50) at WorldView.cpp:1584
#4  0x00000000005810b6 in WorldView::Update (this=0x200f2e50) at WorldView.cpp:1097
#5  0x00000000004d6181 in Pi::MainLoop () at Pi.cpp:1235
#6  0x00000000004d740f in Pi::Start () at Pi.cpp:1101
#7  0x0000000000410f2d in main (argc=<optimized out>, argv=<optimized out>) at main.cpp:130
(gdb)
```

dodomorandi commented:
@impaktor Nice! My fault: I did not try the hyperjump. Fixed with the last commit. Give it a look and let me know if this could break other things (probably not, but it is better to be sure). A question: do you think that the planner should be resetted after a hyperjump? Leaving as it is does not make so much sense... ;)
@nozmajner Woo, thank you! :D If everything goes well, this branch could be merged and I will open another PR for the icons and I will put them in! What do you think about an option inside the settings menu to toggle icons on and off?

impaktor commented:
I rely heavily on you to review and test your own code, so please test various situations. I just tested huperjumping since that is a classic bug trigger when working with system info, since that changes after the jump.

I'd like @nozmajner's hourglass/clock icon to go into _this_ PR, or else we add a new icon that will immediately be replaced in the next PR, which seems a bit silly. My plan was to fix it myself at merge time yesterday, but then that didn't happen due to the segfault problem. Other icons can, as you say, go into another PR.

I don't plan to be at a pioneer-capable computer until next Monday. I _could_ merge before then, but then I need to know that this PR is tested. Someone else can also merge this, before next week.

> A question: do you think that the planner should be resetted after a hyperjump?

Possibly. What if you're viewing the system you are about to jump to? Or is it always on your current system?

dodomorandi commented:
About the bug and debugging... Yes, I know that I have to do some tests. ;) Indeed, it is my fault i did not test hyperjump.

I will push a new commit with the new clock icon, no problem. I will also change the behaviour of the planner, so it is resetted after the hyperjump. The planned orbit is only available starting from the position obtained by current orbit, therefore it is not visible in a new system until jumped.

However, jumping around and trying to make nice maneuvers made me discover one bug, which exists even in the master branch. It is related to the calculation of the position on the orbit using time controls. With the new planning system is easier to detect it, and for this reason I do not know if I should try to fix it (but I do not know how much time I would need) or try to fix the issue in another PR.
What do you think? Meanwhile I will make some changes about icons and planning reset.

dvolk commented on src/SystemView.cpp:
```diff
@@ -28,26 +30,107 @@ static const float WHEEL_SENSITIVITY = .1f;		// Should be a variable in user set
+    else if(absT < 86400)
+        formattedTime << t / 3600. << "h";
+    else
+        formattedTime << t / 86400. << "y";
```
Should be 'd'?

dodomorandi commented:
@dvolk Thank you! Fixed.

laarmen commented:
Hey, don't beat yourself up too badly, we've been known to merge buggy code
ourselves.

impaktor commented:
Oh, yes, it's very common for us to find bugs in each others PRs, or in master the days/weeks after new code is merged. It's the price of progress, and not any major deal.

I'm just letting you know that I don't intend to brood over every line of code in this PR, especially since you seem to know what you're doing. I'll most likely check behaviour and glance at the code a bit.

dodomorandi commented:
Thank you guys, it is just that I hate leaving stupid bugs around ;)

However, I found the reason for the strange behaviour I was talking before: it is a double overflow, which happens when working with dates in the future. It must resolved in a different PR because lots of things need do be changed in any case. I can open an issue, and I can try to give a look to the problem, if you want...

impaktor commented:
Merged manually. I had to cherry pick the ones I wanted, and squashed the bug fixes and different versions of the same icon, to keep our commit history clean.

I did notice that when the planner returns "nan", for instance when playing with it while docked at an orbital station, the "nan" message overlaps with the camera text ("front view", "sidereal view").

dodomorandi commented:
Good work! I created a PR for the issue at https://github.com/pioneerspacesim/pioneer/pull/3437.

If you have any advice to make the commit history cleaner, it is appreciated. I am sorry to let you cherry picking and squashing commits around, but I was worried about readability of changes.

impaktor commented on src/SystemView.cpp:
```diff
@@ -28,26 +30,106 @@ static const float WHEEL_SENSITIVITY = .1f;		// Should be a variable in user set
+		   m_dvNormal   * pNormal.Normalized() +
+		   m_dvRadial   * m_position.Normalized();
+}
+
```
Maybe indentation here is a bit strange. mix of tab and spaces. Nothing worth fixing now, but to keep in mind.

Also, I _think_ we prefer to have opening and closing curly `{...}` on the same colum. @fluffyfreak scream if I'm wrong? Difficult to know when the existing pioneer code is pretty mixed. I think we had something on our wiki on coding style, but cant find it now.

dodomorandi replied:
I got some troubles because I did not get exactly how to format the code ;)
I found different approaches for curly braces, and... I just choose what I thought it was the "usual" way (even if I generally use braces on the same column as well).
What do you think about fixing some rules to use with astyle and making a huge commit with code format refactory?

impaktor commented:
...making a huge commit with code format refactory?

Nah, I wouldn't bother. If we keep to some pre-defined style, with time the code will drift towards uniform correctness by itself, as code gets re-written.

dodomorandi commented:
:+1: Shall we define some astyle rules to make the formatting process easier? Obviously, writing everything on the wiki.

impaktor commented:
http://pioneerwiki.com/wiki/Code_style

dodomorandi commented:
I already read it ;) However, brackets style should be defined (allman or whatever) and other minor things (reference and pointer alignments, max statements length and so on).

impaktor commented:
> brackets style should be defined 

Yes. This is something to be discussed at the forum I think. Feel free to start a thread on it. 

I _think_ most functions have opening and closing brackets on the same column, but for `if` and `for` blocks it varies. It depends who coded the specific source file in question. Perhaps some nifty code style analyzer can give us some statistics on what the heck we're using, or plain regexp search.
